### PR TITLE
RS1029: Do not use reserved diagnostic IDs

### LINKED ISSUE #1727: Write a meta-analyzer to flag analyzers that use VB/C# compiler reserved diagnostic IDs
See https://github.com/dotnet/roslyn/issues/25748 for more details and previous attempt to put such an diagnostic in the compiler itself

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #1727

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx [hunk 1]
```diff
@@ -389,4 +389,13 @@
   <data name="ProvideCustomTagsInDescriptorTitle" xml:space="preserve">
     <value>Provide non-null 'customTags' value to diagnostic descriptor constructor.</value>
   </data>
+  <data name="DoNotUseReservedDiagnosticIdDescription" xml:space="preserve">
```
Comment (Evangelink) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx]: Not sure this is helpful to have a description.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs [hunk 2]
```diff
@@ -658,5 +690,29 @@ private static bool TryParseIdRangeEntry(string entry, out string prefix, out in
 
             return prefix.Length > 0 && suffixStr.Length > 0 && int.TryParse(suffixStr, out suffix);
         }
+
+        private static bool IsReservedDiagnosticId(string ruleId, string assemblyName)
+        {
+            if (ruleId.Length < 6)
```
Comment (Evangelink) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Do we want to flag cases like `CA123`?

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs [hunk 3]
```diff
@@ -27,6 +27,18 @@ public sealed class DiagnosticDescriptorCreationAnalyzer : DiagnosticAnalyzer
         private const string DiagnosticCategoryAndIdRangeFile = "DiagnosticCategoryAndIdRanges.txt";
         private static readonly (string? prefix, int start, int end) s_defaultAllowedIdsInfo = (null, -1, -1);
 
+        private static readonly ImmutableHashSet<string> CADiagnosticIdAllowedAssemblies = ImmutableHashSet.Create(
+            StringComparer.Ordinal,
+            "Microsoft.CodeQuality.Analyzers",
```
Comment (mavasani) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Actually there is one more to be added to this list: https://github.com/dotnet/roslyn-analyzers/blob/master/nuget/Microsoft.CodeAnalysis.VersionCheckAnalyzer/Microsoft.CodeAnalysis.VersionCheckAnalyzer.Package.csproj#L16

https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.VersionCheckAnalyzer/VersionCheckAnalyzer.cs#L16

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs [hunk 4]
```diff
@@ -139,6 +155,16 @@ public sealed class DiagnosticDescriptorCreationAnalyzer : DiagnosticAnalyzer
             description: s_localizableProvideCustomTagsDescription,
             customTags: WellKnownDiagnosticTags.Telemetry);
 
+        public static readonly DiagnosticDescriptor DoNotUseReservedDiagnosticIdRule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotUseReservedDiagnosticIdRuleId,
+            s_localizableDoNotUseReservedDiagnosticIdTitle,
+            s_localizableDoNotUseReservedDiagnosticIdMessage,
+            DiagnosticCategory.MicrosoftCodeAnalysisDesign,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
```
Comment (mavasani) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: I think we all wanted this to be an `Error` severity by default.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx [hunk 5]
```diff
@@ -389,4 +389,13 @@
...
+    <value>DiagnosticId for analyzers should not use reserved IDs.</value>
+  </data>
+  <data name="DoNotUseReservedDiagnosticIdMessage" xml:space="preserve">
+    <value>The DiagnosticId '{0}' is a reserved ID.</value>
```
Comment (mavasani) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx]:
```suggestion
    <value'{0}' is a reserved diagnostic ID.</value>
```

File: src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx [hunk 6]
```diff
@@ -389,4 +389,13 @@
...
+  </data>
+  <data name="DoNotUseReservedDiagnosticIdTitle" xml:space="preserve">
+    <value>DiagnosticId is reserved.</value>
```
Comment (mavasani) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/CodeAnalysisDiagnosticsResources.resx]:
```suggestion
    <value>Do not use reserved diagnostic IDs.</value>
```

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 3 above)
Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Just for additional sanity, can you locally build a NuGet package with this analyzer (execute `eng\common\cibuild.cmd` to generate `Microsoft.CodeAnalysis.Analyzers.nupkg` in artifacts directory) and verify that the new rule does not fire on a project with reference to FxCop analyzers package?

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 2 above)
Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: I think we want to flag even these cases - emphasis is on avoiding non-Microsoft packages having analyzers with CA, BC and CS prefixes.

**Comment by mavasani** [comment]
Tagging @sharwell and @terrajobst for review.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 3 above)
Comment (Evangelink) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: That was a good idea! We missed `Text.Analyzers`. I have also added the C# and VB.NET projects even if they don't have any ID yet.

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Actually with https://github.com/dotnet/roslyn-analyzers/pull/3139, we will also have `Microsoft.CodeAnalysis.NetAnalyzers` and its C# and VB versions. Can you add those as well or at least file a tracking issue to add them?

Comment (Evangelink) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Done

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs [hunk 7]
```diff
@@ -139,6 +162,16 @@ public sealed class DiagnosticDescriptorCreationAnalyzer : DiagnosticAnalyzer
             description: s_localizableProvideCustomTagsDescription,
             customTags: WellKnownDiagnosticTags.Telemetry);
 
+        public static readonly DiagnosticDescriptor DoNotUseReservedDiagnosticIdRule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotUseReservedDiagnosticIdRuleId,
+            s_localizableDoNotUseReservedDiagnosticIdTitle,
+            s_localizableDoNotUseReservedDiagnosticIdMessage,
+            DiagnosticCategory.MicrosoftCodeAnalysisDesign,
+            DiagnosticSeverity.Error,
```
Comment (sharwell) [inline src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: ❗️ Diagnostic analyzers cannot have default severity error. It breaks the analyzer driver's suppression feature. Users who want to see analyzer diagnostics as errors must configure the severity through .editorconfig, a rule set file, and/or set warnings as errors during the build.

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: > No diagnostic analyzers can have default severity error

I don't believe that is true here. We even talked about having this diagnostic in the compiler itself, and settled as having this an analyzer diagnostic with severity error. Tagging @terrajobst

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 4 above)
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: Diagnostic analyzers cannot use Error as a default severity. I filed #3159 to fix this.

File: src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs (hunk 7 above)
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: ➡️ Filed #3159 to fix this

Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: @mavasani The limitation here is not specific to the analyzer in question. Setting a default severity of Error breaks a core compiler feature designed to support scenarios where a team finds an analyzer false positive.

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DiagnosticDescriptorCreationAnalyzer.cs]: I don't understand - analyzer errors can be suppressed just like analyzer warning. Default severity error just ensures a higher visibility for analyzer authors using reserved diagnostic IDs. Additionally, I think we want analyzer authors to absolutely stop using reserved diagnostic IDs

**Pull request merged**